In [111]:
import pandas as pd
import numpy as np

In [ ]:
# %pip install dataframe-image

In [183]:
import dataframe_image as dfi

In [112]:
insurance_df = pd.read_excel('Motor RP Analyst Interview Activity.xlsx', sheet_name='Data')
insurance_df.head()

id start_date  exposure  net_earned_premium  net_written_premium  \
0  16042 2019-02-28  0.002740             0.00000               0.0000   
1  41414 2019-03-20  1.000000           238.33650             238.3365   
2  58837 2019-11-29  0.420765            77.31347             183.7450   
3  65415 2019-02-17  1.000000           494.98150             246.5586   
4  81565 2018-04-24  1.000000           120.05060             120.0506   

   claim_count  is_capped_incurred  incurred  capped_incurred  bi_incurred  \
0            0                 NaN       NaN              NaN          NaN   
1            0                 NaN       NaN              NaN          NaN   
2            0                 NaN       NaN              NaN          NaN   
3            0                 NaN       NaN              NaN          NaN   
4            0                 NaN       NaN              NaN          NaN   

   ...  vehicle_make  d1_age  d1_residency_years  d1_is_uk_resident  \
0  ...          FIAT      20                  20                  1   
1  ...          FORD      25                  25                  1   
2  ...       PEUGEOT      42                  42                  1   
3  ...        NISSAN      31                  31                  1   
4  ...           KIA      61                  61                  1   

   transmission         region  gender  vehicle_purchase_date  \
0        MANUAL  East Midlands    Male    2019-02-01 00:00:00   
1        MANUAL     South East  Female    2017-11-01 00:00:00   
2        MANUAL  West Midlands  Female    2016-09-01 00:00:00   
3     AUTOMATIC     North West  Female    2019-02-16 00:00:00   
4        MANUAL     South West    Male    2013-06-01 00:00:00   

   licence_test_date primary_car_colour  
0         2016-11-26               Blue  
1         2012-03-12               Grey  
2         1994-11-09             Silver  
3         2019-07-01              White  
4         1974-04-07             Silver  

[5 rows x 40 columns]

## Quick Checks

In [113]:
insurance_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 40 columns):
 #   Column                  Non-Null Count   Dtype         
---  ------                  --------------   -----         
 0   id                      100000 non-null  int64         
 1   start_date              100000 non-null  datetime64[ns]
 2   exposure                100000 non-null  float64       
 3   net_earned_premium      99978 non-null   float64       
 4   net_written_premium     100000 non-null  float64       
 5   claim_count             100000 non-null  int64         
 6   is_capped_incurred      7419 non-null    float64       
 7   incurred                7419 non-null    float64       
 8   capped_incurred         7419 non-null    float64       
 9   bi_incurred             7419 non-null    float64       
 10  ad_incurred             7419 non-null    float64       
 11  pd_incurred             7419 non-null    float64       
 12  th_incurred             7419 no

In [114]:
insurance_df.columns

Index(['id', 'start_date', 'exposure', 'net_earned_premium',
       'net_written_premium', 'claim_count', 'is_capped_incurred', 'incurred',
       'capped_incurred', 'bi_incurred', 'ad_incurred', 'pd_incurred',
       'th_incurred', 'ws_incurred', 'bi_count', 'ad_count', 'pd_count',
       'th_count', 'ws_count', 'business_provider', 'cover', 'voluntary_xs',
       'ncd', 'pncd', 'vehicle_use', 'entitlement', 'overnight_location',
       'vehicle_value', 'vehicle_annual_mileage', 'vehicle_age',
       'vehicle_make', 'd1_age', 'd1_residency_years', 'd1_is_uk_resident',
       'transmission', 'region', 'gender', 'vehicle_purchase_date',
       'licence_test_date', 'primary_car_colour'],
      dtype='object')

In [115]:
# Show all cols
insurance_df.describe()

id                     start_date       exposure  \
count  100000.000000                         100000  100000.000000   
mean    50000.500000  2019-03-20 00:31:23.520000256       0.719595   
min         1.000000            2018-04-01 00:00:00       0.002732   
25%     25000.750000            2018-09-15 00:00:00       0.439891   
50%     50000.500000            2019-03-21 00:00:00       0.937159   
75%     75000.250000            2019-09-09 00:00:00       1.000000   
max    100000.000000            2020-03-31 00:00:00       1.000000   
std     28867.657797                            NaN       0.344350   

       net_earned_premium  net_written_premium    claim_count  \
count        99978.000000        100000.000000  100000.000000   
mean           272.432194           364.273773       0.079340   
min          -1134.753000             0.000000       0.000000   
25%            114.164475           181.881875       0.000000   
50%            206.621700           284.771550       0.000000   
75%            359.323175           455.693300       0.000000   
max           3912.240000          4394.661000       4.000000   
std            250.834453           281.664104       0.289941   

       is_capped_incurred       incurred  capped_incurred    bi_incurred  ...  \
count         7419.000000    7419.000000      7419.000000    7419.000000  ...   
mean             0.001483    2797.004851      2589.680535     989.259222  ...   
min              0.000000   -1854.640000     -1854.640000       0.000000  ...   
25%              0.000000      34.500000        34.500000       0.000000  ...   
50%              0.000000     159.500000       159.500000       0.000000  ...   
75%              0.000000    2868.070000      2868.070000       0.000000  ...   
max              1.000000  489297.500000     55259.600000  478339.300000  ...   
std              0.038480   10651.871541      5338.339710    9691.253028  ...   

          ws_count   voluntary_xs            ncd  vehicle_value  \
count  7419.000000  100000.000000  100000.000000  100000.000000   
mean      0.256638     197.238500       5.449590    5525.167150   
min       0.000000       0.000000       0.000000       1.000000   
25%       0.000000     100.000000       2.000000    1695.000000   
50%       0.000000     250.000000       6.000000    3745.000000   
75%       1.000000     250.000000       9.000000    7100.000000   
max       3.000000     600.000000       9.000000  163995.000000   
std       0.447479     136.849868       3.637053    5765.353771   

       vehicle_annual_mileage    vehicle_age         d1_age  \
count           100000.000000  100000.000000  100000.000000   
mean              6878.596280       8.378470      39.937510   
min                  0.000000      -1.000000      17.000000   
25%               5000.000000       5.000000      26.000000   
50%               6000.000000       8.000000      38.000000   
75%               9000.000000      12.000000      51.000000   
max             400000.000000      41.000000     118.000000   
std               4056.018302       4.680995      15.415234   

       d1_residency_years  d1_is_uk_resident              licence_test_date  
count       100000.000000       100000.00000                          99993  
mean            38.400150            0.92853  2002-08-24 07:05:07.065494656  
min              0.000000            0.00000            1958-03-04 00:00:00  
25%             25.000000            1.00000            1994-05-24 00:00:00  
50%             36.000000            1.00000            2003-05-02 00:00:00  
75%             51.000000            1.00000            2013-03-29 00:00:00  
max            118.000000            1.00000            2020-09-30 00:00:00  
std             16.449685            0.25761                            NaN  

[8 rows x 28 columns]

In [116]:
## Check unique values in objects

def show_object_uniques(df):
    """
    Display all unique values for object (string/categorical) columns in a DataFrame.
    """
    for col in df.select_dtypes(include=['object']).columns:
        print(f"\nColumn: {col}")
        print("-" * (len(col) + 9))
        print(df[col].unique())
        

In [117]:
## Check unique values in objects
show_object_uniques(insurance_df)


Column: business_provider
--------------------------
['CTMDefaultCampaign' 'HDWebsiteDefaultCampaign' 'MSMDefaultCampaign'
 'QuotezoneDefaultCampaign' 'ConfusedDefaultCampaign'
 'GoCompareDefaultCampaign' 'Confused_ECI_IP' 'Unknown'
 'DefaultDefaultCampaign' 'PPC-BingDefaultCampaign' 'CTM_ECI_PC' 'CTM_ECI'
 'CTM_ECI_IP' 'TopCashBackDefaultCampaign'
 'Google_Brand-Gk-Hastings_Direct' 'Confused_ECI' 'CTM_ECI_Hastings'
 'InsurerDefaultCampaign' 'Google_Brand-Gk-Hastings_Direct-Mobile'
 'Confused_ECI_PC' 'uSwitchDefaultCampaign' 'Bing_Car_Brand'
 'QuidcoDefaultCampaign' 'ContactUsPage' 'Confused_ECI_Hastings'
 'IPWebsiteDefaultCampaign' 'PPC-GoogleDefaultCampaign'
 'Google_Car_Gdn/Sr' 'HDWebsiteMulticarTestOne'
 'Google_Generic-Car-Car_Type' 'Google_Car_Generic'
 'Google_Brand-Car_Insurance' 'Google_Car_Brand'
 'Google_Brand-Mobile-Van_Insurance-Exact' 'RenewalRebrokeDefaultCampaign'
 'Google_Brand-Remarketing-Brand' 'PCWebsiteDefaultCampaign'
 'MTARebrokeDefaultCampaign' 'Google_Brand-Gk

In [118]:
insurance_df.isnull().sum()

id                            0
start_date                    0
exposure                      0
net_earned_premium           22
net_written_premium           0
claim_count                   0
is_capped_incurred        92581
incurred                  92581
capped_incurred           92581
bi_incurred               92581
ad_incurred               92581
pd_incurred               92581
th_incurred               92581
ws_incurred               92581
bi_count                  92581
ad_count                  92581
pd_count                  92581
th_count                  92581
ws_count                  92581
business_provider             0
cover                         0
voluntary_xs                  0
ncd                           0
pncd                          0
vehicle_use                   0
entitlement                   0
overnight_location            0
vehicle_value                 0
vehicle_annual_mileage        0
vehicle_age                   0
vehicle_make                776
d1_age  

In [119]:
# Check cols with null net_earned_premium
insurance_df[insurance_df['net_earned_premium'].isnull()].head()

id start_date  exposure  net_earned_premium  net_written_premium  \
12403  39729 2019-12-02  0.412568                 NaN                  0.0   
28675  34290 2018-12-01  0.564384                 NaN                  0.0   
29357  87570 2018-11-14  0.608219                 NaN                  0.0   
30700  80529 2018-09-13  1.000000                 NaN                  0.0   
31197  44999 2018-12-20  0.512329                 NaN                  0.0   

       claim_count  is_capped_incurred  incurred  capped_incurred  \
12403            0                 NaN       NaN              NaN   
28675            0                 NaN       NaN              NaN   
29357            0                 NaN       NaN              NaN   
30700            0                 NaN       NaN              NaN   
31197            0                 NaN       NaN              NaN   

       bi_incurred  ...  vehicle_make  d1_age  d1_residency_years  \
12403          NaN  ...          FIAT      24                  24   
28675          NaN  ...          FORD      28                  28   
29357          NaN  ...       CITROEN      45                  45   
30700          NaN  ...           BMW      38                  29   
31197          NaN  ...          FORD      50                  50   

       d1_is_uk_resident  transmission         region  gender  \
12403                  1        MANUAL     South West  Female   
28675                  1        MANUAL       Scotland  Female   
29357                  1        MANUAL     North West    Male   
30700                  0        MANUAL  East Midlands    Male   
31197                  1        MANUAL  West Midlands    Male   

       vehicle_purchase_date  licence_test_date primary_car_colour  
12403    2019-03-01 00:00:00         2018-12-01              White  
28675    2017-07-01 00:00:00         2017-03-17              Black  
29357    2018-11-13 00:00:00         1998-11-13              Black  
30700    2012-09-01 00:00:00         1998-08-30               Grey  
31197    2018-11-01 00:00:00         2014-02-10             Silver  

[5 rows x 40 columns]

In [120]:
## Checking neagitve inccured values
insurance_df[insurance_df['incurred'] < 0].head()

id start_date  exposure  net_earned_premium  net_written_premium  \
90   70381 2018-05-01  1.000000            218.1073             218.1073   
180  99011 2018-05-07  1.000000            267.2897             271.3575   
487   5840 2018-04-07  1.000000            280.8841             237.0196   
545   1677 2019-01-26  0.726027            650.7096             899.6513   
590  45603 2019-02-06  1.000000            302.1702             302.1702   

     claim_count  is_capped_incurred  incurred  capped_incurred  bi_incurred  \
90             1                 0.0     -61.0            -61.0          0.0   
180            1                 0.0    -192.0           -192.0          0.0   
487            1                 0.0     -84.5            -84.5          0.0   
545            1                 0.0      -5.0             -5.0          0.0   
590            1                 0.0     -70.5            -70.5          0.0   

     ...  vehicle_make  d1_age  d1_residency_years  d1_is_uk_resident  \
90   ...          FORD      40                  40                  1   
180  ...           BMW      62                  62                  1   
487  ...        NISSAN      31                  31                  1   
545  ...      VAUXHALL      23                  23                  1   
590  ...        NISSAN      31                  31                  1   

     transmission         region  gender  vehicle_purchase_date  \
90         MANUAL  East Midlands  Female    2018-04-16 00:00:00   
180        MANUAL     North West  Female    2016-02-01 00:00:00   
487        MANUAL     North West    Male    2018-04-01 00:00:00   
545        MANUAL  East Midlands    Male    2018-08-01 00:00:00   
590        MANUAL     South East  Female    2019-02-01 00:00:00   

     licence_test_date primary_car_colour  
90          2009-04-16               Blue  
180         1998-04-25              White  
487         2005-04-06                NaN  
545         2014-01-24                NaN  
590         2005-02-09                NaN  

[5 rows x 40 columns]

In [121]:
# Check vehicle age with negative?? # Include all cols
pd.set_option("display.max_columns", None)
result = insurance_df[insurance_df['vehicle_age'] < 0].head(1)
# Reset option
print(result)
pd.reset_option("display.max_columns")

        id start_date  exposure  net_earned_premium  net_written_premium  \
363  97981 2018-09-12       1.0           1292.4962            1256.8816   

     claim_count  is_capped_incurred  incurred  capped_incurred  bi_incurred  \
363            1                 0.0   7917.35          7917.35          0.0   

     ad_incurred  pd_incurred  th_incurred  ws_incurred  bi_count  ad_count  \
363      2863.04      5054.31          0.0          0.0       0.0       1.0   

     pd_count  th_count  ws_count business_provider cover  voluntary_xs  ncd  \
363       1.0       0.0       0.0      Confused_ECI  COMP           500    0   

    pncd vehicle_use entitlement overnight_location  vehicle_value  \
363    N           S         IAC                  R          20000   

     vehicle_annual_mileage  vehicle_age vehicle_make  d1_age  \
363                    2000           -1          BMW      22   

     d1_residency_years  d1_is_uk_resident transmission  region gender  \
363                 

### Notes
- Data dictionary mentions 42 columns while the data tab only has 40 columns.

### Null Values
Columns that contain null value:
- primary_car_colour
- region
- transmission
- vehicle_make
- net_earned_premium
    
- is_capped_incurred links to -> incurred, capped_incurred, bi_incurred, ad_incurred, pd_incurred, th_incurred, ws_incurred          
    bi_count, ad_count, pd_count, th_count, ws_count

### Unusual Values
    - Some vehicle ages are -1 ?
    - A vehicle value of 1 ?
    - Annual vehicle mileage of 400000 ?
    - d1 age and residency years of 118 ?
    - earned premiun and incurred at negative?


## Cleaning Missing/Unusual Values

In [122]:
insurance_df_clean = insurance_df.copy()

In [205]:

def null_percentages(df):
    """
    Returns the percentage of missing values per column in a DataFrame,
    excluding columns with no nulls.
    """
    null_counts = df.isnull().sum()
    total_rows = len(df)
    null_percent = (null_counts / total_rows) * 100
    
    # Filter out columns with 0 nulls
    null_df = pd.DataFrame({
        'Null %': null_percent.round(2)
    })
    null_df = null_df[null_df['Null %'] > 0]

    return null_df.sort_values(by='Null %', ascending=False)

null_percentages(insurance_df)

Null %
is_capped_incurred   92.58
incurred             92.58
capped_incurred      92.58
pd_count             92.58
bi_incurred          92.58
ad_incurred          92.58
pd_incurred          92.58
th_incurred          92.58
ws_incurred          92.58
bi_count             92.58
ad_count             92.58
ws_count             92.58
th_count             92.58
primary_car_colour   16.26
vehicle_make          0.78
transmission          0.78
region                0.14
net_earned_premium    0.02
licence_test_date     0.01

In [124]:
insurance_df_clean['is_capped_incurred'].unique()

array([nan,  0.,  1.])

In [125]:
incurred_cols = [
    "is_capped_incurred", "incurred", "capped_incurred",
    "pd_count", "bi_incurred", "ad_incurred", "pd_incurred",
    "th_incurred", "ws_incurred", "bi_count", "ad_count",
    "ws_count", "th_count"
]

In [126]:
for col in incurred_cols:
    insurance_df_clean[col] = insurance_df_clean[col].apply(lambda x: 0 if pd.isna(x) else x).astype(int)


In [127]:
insurance_df_clean['primary_car_colour'].unique()

array(['Blue', 'Grey', 'Silver', 'White', 'Black', 'Red', 'Beige', nan,
       'Green', 'Purple', 'Orange', 'Gold', 'Brown', 'Turquoise',
       'Bronze', 'Yellow', 'Maroon', 'Pink', 'Multi-colour', 'Cream'],
      dtype=object)

In [128]:
insurance_df_clean['primary_car_colour'] = insurance_df_clean['primary_car_colour'].apply(
    lambda x: 'Missing' if pd.isna(x) else x
)

In [129]:
insurance_df_clean['vehicle_make'].unique()

array(['FIAT', 'FORD', 'PEUGEOT', 'NISSAN', 'KIA', 'VAUXHALL',
       'VOLKSWAGEN', 'TOYOTA', 'RENAULT', 'SKODA', 'CITROEN', 'HYUNDAI',
       'ALFA ROMEO', 'HONDA', 'BMW', 'AUDI', 'MERCEDES-BENZ', 'VOLVO',
       'CHRYSLER', nan, 'SMART', 'MINI', 'MAZDA', 'SSANGYONG', 'SEAT',
       'SUZUKI', 'SAAB', 'JAGUAR', 'MITSUBISHI', 'PORSCHE', 'SUBARU',
       'DACIA', 'CHEVROLET', 'LEXUS', 'DAEWOO', 'DODGE', 'MG-MOTOR UK',
       'LANDROVER', 'DS', 'ROVER', 'BMW320D', 'INFINITI', 'JEEP', 'MG',
       'PERODUA', 'DAIHATSU', 'DATSUN', 'AUSTIN', 'CADILLAC', 'PROTON',
       'FIAT500', 'JAGUARXE', 'ISUZU', 'MG-MOTOR UKZS', 'LOTUS',
       'TOYOTACOROLLA', 'OPEL', 'LEXUSUX', 'FIAT500C', 'AUDIQ5',
       'BMW320I', 'JAGUARF-PACE', 'TOYOTARAV-4', 'NAC MG', 'DAIMLER',
       'TESLA', 'FIATPANDA', 'FIAT500X'], dtype=object)

In [130]:
insurance_df_clean['vehicle_make'] = insurance_df_clean['vehicle_make'].apply(
    lambda x: 'Missing' if pd.isna(x) else x
)

In [131]:
insurance_df_clean['transmission'].unique()

array(['MANUAL', 'AUTOMATIC', nan], dtype=object)

In [132]:
insurance_df_clean['transmission'] = insurance_df_clean['transmission'].apply(
    lambda x: 'Missing' if pd.isna(x) else x
)

In [133]:
insurance_df_clean['region'].unique()

array(['East Midlands', 'South East', 'West Midlands', 'North West',
       'South West', 'London', 'Scotland', 'North', 'Yorkshire', 'Wales',
       'East Anglia', 'Outer London', 'N Ireland', nan], dtype=object)

In [134]:
insurance_df_clean['region'] = insurance_df_clean['region'].apply(
    lambda x: 'Missing' if pd.isna(x) else x
)

In [135]:
# Check nan net earned premium
insurance_df_clean[insurance_df_clean['net_earned_premium'].isna()]

id start_date  exposure  net_earned_premium  net_written_premium  \
12403  39729 2019-12-02  0.412568                 NaN                  0.0   
28675  34290 2018-12-01  0.564384                 NaN                  0.0   
29357  87570 2018-11-14  0.608219                 NaN                  0.0   
30700  80529 2018-09-13  1.000000                 NaN                  0.0   
31197  44999 2018-12-20  0.512329                 NaN                  0.0   
33518  34292 2018-09-09  0.786301                 NaN                  0.0   
34562  98830 2019-08-04  0.740437                 NaN                  0.0   
38848  83779 2019-06-10  0.890710                 NaN                  0.0   
40022  59739 2018-11-27  1.000000                 NaN                  0.0   
41601  46202 2020-01-17  0.286885                 NaN                  0.0   
47632  87274 2019-04-16  0.185792                 NaN                  0.0   
54390  72190 2018-11-06  0.794520                 NaN                  0.0   
58853  25137 2019-04-12  1.000000                 NaN                  0.0   
71409   9023 2020-02-28  0.172131                 NaN                  0.0   
76530  56353 2019-04-26  0.158470                 NaN                  0.0   
77736  29725 2019-08-27  0.677596                 NaN                  0.0   
81692  23856 2019-08-31  0.666667                 NaN                  0.0   
83492  21738 2019-04-06  0.213115                 NaN                  0.0   
85134  56839 2018-10-01  1.000000                 NaN                  0.0   
87873  34639 2019-05-07  0.128415                 NaN                  0.0   
91396  62306 2019-07-08  0.125683                 NaN                  0.0   
92722  76037 2018-10-25  0.660274                 NaN                  0.0   

       claim_count  is_capped_incurred  incurred  capped_incurred  \
12403            0                   0         0                0   
28675            0                   0         0                0   
29357            0                   0         0                0   
30700            0                   0         0                0   
31197            0                   0         0                0   
33518            0                   0         0                0   
34562            0                   0         0                0   
38848            0                   0         0                0   
40022            1                   0       114              114   
41601            0                   0         0                0   
47632            0                   0         0                0   
54390            0                   0         0                0   
58853            0                   0         0                0   
71409            0                   0         0                0   
76530            0                   0         0                0   
77736            0                   0         0                0   
81692            0                   0         0                0   
83492            0                   0         0                0   
85134            1                   0       230              230   
87873            0                   0         0                0   
91396            0                   0         0                0   
92722            0                   0         0                0   

       bi_incurred  ...   vehicle_make  d1_age  d1_residency_years  \
12403            0  ...           FIAT      24                  24   
28675            0  ...           FORD      28                  28   
29357            0  ...        CITROEN      45                  45   
30700            0  ...            BMW      38                  29   
31197            0  ...           FORD      50                  50   
33518            0  ...        RENAULT      39                  39   
34562            0  ...       VAUXHALL      25                  25   
38848            0  ...           MINI      21                  21 

In [136]:
insurance_df_clean['premium_flag'] = np.where(
    insurance_df_clean['net_earned_premium'].isna(), 'Missing Premium', 'Normal'
)

### Notes
- Around 22 rows have missing premium but some still have exposure/claims. These could represent cancellations or system issues, and I flagged them separately rather than forcing values.

In [137]:
insurance_df_clean = insurance_df_clean.dropna(subset=['licence_test_date'])

In [138]:
null_percentages(insurance_df_clean)

Null Count  Null %
net_earned_premium          22    0.02

In [139]:
insurance_df_clean.head(3)

id start_date  exposure  net_earned_premium  net_written_premium  \
0  16042 2019-02-28  0.002740             0.00000               0.0000   
1  41414 2019-03-20  1.000000           238.33650             238.3365   
2  58837 2019-11-29  0.420765            77.31347             183.7450   

   claim_count  is_capped_incurred  incurred  capped_incurred  bi_incurred  \
0            0                   0         0                0            0   
1            0                   0         0                0            0   
2            0                   0         0                0            0   

   ...  d1_age  d1_residency_years  d1_is_uk_resident  transmission  \
0  ...      20                  20                  1        MANUAL   
1  ...      25                  25                  1        MANUAL   
2  ...      42                  42                  1        MANUAL   

          region  gender  vehicle_purchase_date  licence_test_date  \
0  East Midlands    Male    2019-02-01 00:00:00         2016-11-26   
1     South East  Female    2017-11-01 00:00:00         2012-03-12   
2  West Midlands  Female    2016-09-01 00:00:00         1994-11-09   

   primary_car_colour premium_flag  
0                Blue       Normal  
1                Grey       Normal  
2              Silver       Normal  

[3 rows x 41 columns]

In [140]:
insurance_df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 99993 entries, 0 to 99999
Data columns (total 41 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   id                      99993 non-null  int64         
 1   start_date              99993 non-null  datetime64[ns]
 2   exposure                99993 non-null  float64       
 3   net_earned_premium      99971 non-null  float64       
 4   net_written_premium     99993 non-null  float64       
 5   claim_count             99993 non-null  int64         
 6   is_capped_incurred      99993 non-null  int64         
 7   incurred                99993 non-null  int64         
 8   capped_incurred         99993 non-null  int64         
 9   bi_incurred             99993 non-null  int64         
 10  ad_incurred             99993 non-null  int64         
 11  pd_incurred             99993 non-null  int64         
 12  th_incurred             99993 non-null  int64      

### Unusual values

In [141]:
insurance_df_clean.describe()

id                     start_date      exposure  \
count   99993.000000                          99993  99993.000000   
mean    49999.401278  2019-03-20 00:14:20.604242432      0.719620   
min         1.000000            2018-04-01 00:00:00      0.002732   
25%     24999.000000            2018-09-15 00:00:00      0.439891   
50%     49999.000000            2019-03-21 00:00:00      0.937159   
75%     74999.000000            2019-09-09 00:00:00      1.000000   
max    100000.000000            2020-03-31 00:00:00      1.000000   
std     28867.665030                            NaN      0.344343   

       net_earned_premium  net_written_premium   claim_count  \
count        99971.000000         99993.000000  99993.000000   
mean           272.445317           364.284726      0.079346   
min          -1134.753000             0.000000      0.000000   
25%            114.172700           181.885200      0.000000   
50%            206.635100           284.776200      0.000000   
75%            359.361400           455.708000      0.000000   
max           3912.240000          4394.661000      4.000000   
std            250.837938           281.669489      0.289951   

       is_capped_incurred       incurred  capped_incurred    bi_incurred  ...  \
count        99993.000000   99993.000000     99993.000000   99993.000000  ...   
mean             0.000110     207.507266       192.124839      73.397338  ...   
min              0.000000   -1854.000000     -1854.000000       0.000000  ...   
25%              0.000000       0.000000         0.000000       0.000000  ...   
50%              0.000000       0.000000         0.000000       0.000000  ...   
75%              0.000000       0.000000         0.000000       0.000000  ...   
max              1.000000  489297.000000     55259.000000  478339.000000  ...   
std              0.010488    2992.410115      1604.572206    2652.312513  ...   

           ws_count  voluntary_xs           ncd  vehicle_value  \
count  99993.000000  99993.000000  99993.000000   99993.000000   
mean       0.019041    197.241307      5.449511    5525.261068   
min        0.000000      0.000000      0.000000       1.000000   
25%        0.000000    100.000000      2.000000    1695.000000   
50%        0.000000    250.000000      6.000000    3745.000000   
75%        0.000000    250.000000      9.000000    7100.000000   
max        3.000000    600.000000      9.000000  163995.000000   
std        0.139208    136.851063      3.637080    5765.489110   

       vehicle_annual_mileage   vehicle_age        d1_age  d1_residency_years  \
count            99993.000000  99993.000000  99993.000000        99993.000000   
mean              6878.517776      8.378396     39.937706           38.400398   
min                  0.000000     -1.000000     17.000000            0.000000   
25%               5000.000000      5.000000     26.000000           25.000000   
50%               6000.000000      8.000000     38.000000           36.000000   
75%               9000.000000     12.000000     51.000000           51.000000   
max             400000.000000     41.000000    118.000000          118.000000   
std               4056.102573      4.681005     15.415524           16.449851   

       d1_is_uk_resident              licence_test_date  
count       99993.000000                          99993  
mean            0.928535  2002-08-24 07:05:07.065494656  
min             0.000000            1958-03-04 00:00:00  
25%             1.000000            1994-05-24 00:00:00  
50%             1.000000            2003-05-02 00:00:00  
75%             1.000000            2013-03-29 00:00:00  
max             1.000000            2020-09-30 00:00:00  
std             0.257601                            NaN  

[8 rows x 28 columns]

In [142]:
## Vehicles with ages less than 0
insurance_df_clean.loc[insurance_df_clean['vehicle_age'] < 0, 'vehicle_age'].unique()

array([-1])

In [143]:
## Vehicle unusual values
insurance_df_clean.loc[insurance_df_clean['vehicle_value'] < 100, 'vehicle_value'].unique()

array([12, 95,  3,  8,  6,  5, 45, 20, 30,  9, 10,  4,  1, 11, 38])

In [144]:
insurance_df_clean.loc[insurance_df_clean['vehicle_annual_mileage'] > 100000, 'vehicle_annual_mileage'].unique()

array([143000, 106000, 400000, 129999, 120000])

In [145]:
insurance_df_clean.loc[insurance_df_clean['d1_age'] > 100, 'd1_age'].unique()

array([118])

In [146]:
insurance_df_clean.loc[insurance_df_clean['d1_age'] < 18, 'd1_age'].unique()

array([17])

In [147]:
insurance_df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 99993 entries, 0 to 99999
Data columns (total 41 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   id                      99993 non-null  int64         
 1   start_date              99993 non-null  datetime64[ns]
 2   exposure                99993 non-null  float64       
 3   net_earned_premium      99971 non-null  float64       
 4   net_written_premium     99993 non-null  float64       
 5   claim_count             99993 non-null  int64         
 6   is_capped_incurred      99993 non-null  int64         
 7   incurred                99993 non-null  int64         
 8   capped_incurred         99993 non-null  int64         
 9   bi_incurred             99993 non-null  int64         
 10  ad_incurred             99993 non-null  int64         
 11  pd_incurred             99993 non-null  int64         
 12  th_incurred             99993 non-null  int64      

## Export Clean data

### Clean data is only for Tableau use

In [148]:
## Export Clean data
insurance_df_clean.to_csv('insurance_data_clean.csv', index=False)

# Section 1

### Claim Count

In [149]:
# Unique values in claim count including objects
insurance_df['claim_count'].unique()

array([0, 1, 2, 3, 4])

In [150]:
claim_count = (insurance_df['claim_count']).sum()
claim_count

7934

In [151]:
## More than one claim?
claim_count = (insurance_df['claim_count'] > 1).sum()
claim_count

482

### Average value of claim

In [201]:
# Average claim without cap check
average_claim = insurance_df[insurance_df['claim_count'] > 0]['incurred'].mean()
average_claim

2797.0048510580937

In [153]:
summary = (insurance_df.assign(
              start_date=pd.to_datetime(insurance_df['start_date'], errors='coerce'),
              year=lambda d: d['start_date'].dt.year)
           .groupby('year', dropna=True)
           .agg(covers_started=('id','count'),
                claim_count=('claim_count','sum'),
                earned_premium=('net_earned_premium','sum'),
                written_premium=('net_written_premium','sum'))
           .reset_index()
           .rename(columns={'year':'Year',
                            'covers_started':'Covers Started',
                            'claim_count':'Claim Count',
                            'earned_premium':'Earned Premium',
                            'written_premium':'Written Premium'})
           .sort_values('Year')
          )

# Display in £ millions with 2 dp (keep counts as ints)
summary_present = summary.copy()
summary_present['Earned Premium (£m)']  = (summary_present['Earned Premium']/1e6).round(2)
summary_present['Written Premium (£m)'] = (summary_present['Written Premium']/1e6).round(2)
summary_present[['Covers Started','Claim Count']] = summary_present[['Covers Started','Claim Count']].astype(int)

# Final columns (readable)
summary_present = summary_present[['Year','Covers Started','Claim Count',
                                   'Earned Premium (£m)','Written Premium (£m)']]
summary_present


Year  Covers Started  Claim Count  Earned Premium (£m)  \
0  2018           38539         4065                13.37   
1  2019           49808         3700                13.07   
2  2020           11653          169                 0.79   

   Written Premium (£m)  
0                 14.65  
1                 17.70  
2                  4.08

# Biggest Factors

In [154]:
insurance_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 40 columns):
 #   Column                  Non-Null Count   Dtype         
---  ------                  --------------   -----         
 0   id                      100000 non-null  int64         
 1   start_date              100000 non-null  datetime64[ns]
 2   exposure                100000 non-null  float64       
 3   net_earned_premium      99978 non-null   float64       
 4   net_written_premium     100000 non-null  float64       
 5   claim_count             100000 non-null  int64         
 6   is_capped_incurred      7419 non-null    float64       
 7   incurred                7419 non-null    float64       
 8   capped_incurred         7419 non-null    float64       
 9   bi_incurred             7419 non-null    float64       
 10  ad_incurred             7419 non-null    float64       
 11  pd_incurred             7419 non-null    float64       
 12  th_incurred             7419 no

In [179]:
df = insurance_df.copy()
df = df[(df["claim_count"] > 0) & df["capped_incurred"].notna() & (df["capped_incurred"] > 0)]
df["claim_size"] = df["capped_incurred"] / df["claim_count"]

def avg_claim_size_by(factor):
    """
    Calculate average (weighted) and median claim size by a given factor.
    Keeps structure close to your original.
    """
    grouped = (
        df.groupby(factor, dropna=False)
          .agg(
              total_incurred=("capped_incurred", "sum"),
              total_claims=("claim_count", "sum"),
              median_claim_size=("claim_size", "median")  # typical (skew-robust)
          )
    )
    # avoid divide-by-zero & negatives
    grouped = grouped[grouped["total_claims"] > 0]
    grouped["avg_claim_size"] = grouped["total_incurred"] / grouped["total_claims"]
    grouped = grouped[grouped["avg_claim_size"] > 0]
    return grouped.sort_values("avg_claim_size", ascending=False)

factors = [
    "cover","region","d1_age","vehicle_value","vehicle_age","ncd","pncd",
    "overnight_location","vehicle_make","transmission","primary_car_colour",
    "business_provider","voluntary_xs","vehicle_annual_mileage",
    "entitlement","d1_residency_years","d1_is_uk_resident","gender"
]

summary = []

for factor in factors:
    table = avg_claim_size_by(factor)

    min_val = table["avg_claim_size"].min()
    max_val = table["avg_claim_size"].max()
    spread  = max_val - min_val
    overall_median = table["median_claim_size"].median()  # median of category medians

    summary.append({
        "Factor": factor,
        "Min Avg Claim Size": round(min_val, 2),
        "Max Avg Claim Size": round(max_val, 2),
        "Spread": round(spread, 2),
        "Median (overall)": round(overall_median, 2)
    })

summary_df = pd.DataFrame(summary).sort_values("Spread", ascending=False).reset_index(drop=True)

In [190]:
top5 = summary_df.head(5)
top5

Factor  Min Avg Claim Size  Max Avg Claim Size    Spread  \
0           vehicle_value                0.10            49995.00  49994.90   
1  vehicle_annual_mileage               22.75            35166.79  35144.04   
2             vehicle_age             1247.75            12484.15  11236.40   
3            vehicle_make               50.00             8573.29   8523.29   
4       business_provider              130.10             6159.87   6029.77   

   Median (overall)  
0           1056.71  
1            937.20  
2           1315.08  
3            866.43  
4            916.33

# Formating for presentation

In [200]:
present_top5 = summary_df.head(5).copy().rename(columns={
    "Factor": "Factor",
    "Min Avg Claim Size": "Min Avg (£)",
    "Max Avg Claim Size": "Max Avg (£)",
    "Spread": "Spread (£)",
    "Median (overall)": "Median (£)"
})[["Factor", "Min Avg (£)", "Max Avg (£)", "Spread (£)", "Median (£)"]]

# ✨ PRETTIFY factor text
present_top5["Factor"] = (
    present_top5["Factor"]
        .astype(str)
        .str.replace("_", " ")
        .str.title()  
)

money_fmt = "£{:,.1f}"
ACCENT = "#2563eb"

styled = (
    present_top5.style
        .hide(axis="index")
        .format({
            "Min Avg (£)": money_fmt,
            "Max Avg (£)": money_fmt,
            "Spread (£)": money_fmt,
            "Median (£)": money_fmt
        })
        # alignments
        .set_properties(subset=["Factor"], **{"text-align": "left"})
        .set_properties(subset=["Min Avg (£)", "Max Avg (£)", "Median (£)"],
                        **{"text-align": "right"})
        # highlight spread
        .set_properties(subset=["Spread (£)"], **{"text-align": "right",
                                                  "color": ACCENT,
                                                  "font-weight": "700"})
        # 🔴 nicer factor text styling
        .set_properties(subset=["Factor"], **{
            "color": "#d32f2f",              # softer red
            "font-weight": "600",
            "text-decoration": "none",
            "font-family": "Helvetica, Arial, sans-serif"
        })
)

dfi.export(styled, "top5_claim_factors.png", table_conversion="matplotlib", dpi=300)
print("Saved PNG -> top5_claim_factors.png")


Saved PNG -> top5_claim_factors.png


In [206]:
# For section 1 Question 3
null_percentages(insurance_df)

Null %
is_capped_incurred   92.58
incurred             92.58
capped_incurred      92.58
pd_count             92.58
bi_incurred          92.58
ad_incurred          92.58
pd_incurred          92.58
th_incurred          92.58
ws_incurred          92.58
bi_count             92.58
ad_count             92.58
ws_count             92.58
th_count             92.58
primary_car_colour   16.26
vehicle_make          0.78
transmission          0.78
region                0.14
net_earned_premium    0.02
licence_test_date     0.01

In [ ]:
# Mapping of field names to presentable labels
field_name_map = {
    "is_capped_incurred": "Is Capped Incurred",
    "incurred": "Incurred",
    "capped_incurred": "Capped Incurred",
    "bi_incurred": "Bodily Injury Incurred",
    "ad_incurred": "Accidental Damage Incurred",
    "pd_incurred": "Property Damage Incurred",
    "th_incurred": "Theft Incurred",
    "ws_incurred": "Windscreen Incurred",
    "bi_count": "Bodily Injury Count",
    "ad_count": "Accidental Damage Count",
    "pd_count": "Property Damage Count",
    "th_count": "Theft Count",
    "ws_count": "Windscreen Count",
    "primary_car_colour": "Primary Car Colour",
    "vehicle_make": "Vehicle Make",
    "transmission": "Transmission",
    "region": "Region",
    "net_earned_premium": "Net Earned Premium",
    "licence_test_date": "Licence Test Date"
}

# Apply mapping
missing_full = null_percentages(insurance_df).reset_index().rename(
    columns={"index": "Field"}
)
missing_full["Field"] = missing_full["Field"].replace(field_name_map)

# Style and export (percentage only)
PCT_FMT = "{:,.2f}%"
styled_missing = (
    missing_full[["Field", "Missing (%)"]].style
        .hide(axis="index")
        .format({"Missing (%)": PCT_FMT})
        .set_properties(subset=["Field"], **{
            "text-align": "left",
            "color": "#374151",
            "font-weight": "600",
            "font-family": "Helvetica, Arial, sans-serif"
        })
        .set_properties(subset=["Missing (%)"], **{
            "text-align": "right",
            "color": "#2563eb",
            "font-weight": "700"
        })
)

dfi.export(styled_missing, "missing_data_summary.png", table_conversion="matplotlib", dpi=300)
print("Saved PNG -> missing_data_summary.png")


Saved PNG -> missing_data_summary.png
